### PR TITLE
*: REORGANIZE PARTITION (#38535)

### LINKED ISSUE #38535: Support REORGANIZE PARTITION Key Dev Task
# Support Reorganize Partition Dev Task

Related Issues: #15000
Design PR:
- [x] https://github.com/pingcap/tidb/pull/38246

## Dev Tasks
- [x] Parser support
- [x] Copy records, #38460
- [x] Recreate the indexes in the new partitions, #38460
- [x] Double write data in foreground sessions (concurrent DML during DDL), #38508
- [x] Handle `reorganize partition` command in ddl_api.go and ddl.go, #40473
- [x] Error handling and DDL rollback, #40543
- [x] Delete range of old partitions, #40543
- [x] Update Table handling, #40714
- [ ] TiFlash handling, #40715
- [x] Add telemetry for reorganize partition, #40258
- [ ] Support in TiFlash https://github.com/pingcap/tiflash/issues/6444
- [ ] Support in TiCDC https://github.com/pingcap/tiflow/pull/8114

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #38535

Problem Summary:
This is a manual merge of the feature branch 'feature/reorganize-partition' containing #38460, #38508, #40473, #40543, #40714 and #40258.

### What is changed and how it works?
Somehow the feature branch rewrote the commits from master when updating, so this is a manual squash of the content of the feature branch, where each PR has been reviewed and approved.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support ALTER TABLE <table name> REORGANIZE PARTITION <list of partitions> INTO <list of partition definitions>
```

### CODE REVIEW

File: ddl/column.go [hunk 1]
```diff
@@ -1052,9 +1052,35 @@ func BuildElements(changingCol *model.ColumnInfo, changingIdxs []*model.IndexInf
...
+		done := false
+		for !done {
+			p := tbl.GetPartition(reorgInfo.PhysicalTableID)
+			if p == nil {
+				return dbterror.ErrCancelledDDLJob.GenWithStack("Can not find partition id %d for table %d", reorgInfo.PhysicalTableID, t.Meta().ID)
+			}
+			workType := typeReorgPartitionWorker
+			if reorgInfo.Job.Type != model.ActionReorganizePartition {
+				workType = typeUpdateColumnWorker
+				panic("FIXME: See https://github.com/pingcap/tidb/issues/39915")
```
Comment (bb7133) [inline ddl/column.go:1067]: Is it better to return an error instead?

File: ddl/column.go [hunk 2]
```diff
@@ -1085,6 +1111,10 @@ func (w *worker) updateCurrentElement(t table.Table, reorgInfo *reorgInfo) error
 		}
 	}
 
+	if _, ok := t.(table.PartitionedTable); ok {
+		// TODO: Remove this
+		panic("FIXME: this got reverted and needs to be back!")
```
Comment (bb7133) [inline ddl/column.go:1116]: Ditto

File: ddl/column.go (hunk 1 above)
Comment (mjonss) [reply ddl/column.go]: Fixed

File: ddl/column.go (hunk 2 above)
Comment (mjonss) [reply ddl/column.go]: Fixed.

**Review by bb7133 — APPROVED** [review]
LGTM

**Review by tiancaiamao — APPROVED** [review]

**Comment by hawkingrei** [comment]
/hold

CI is too busy.

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**